album.go

```diff
@@ -188,60 +234,162 @@ func (a *Album) GetAllImageKeysFromBucket() ([]string, error) {
+		//note albumOrdering would be empty, error checking matters!
+		return albumOrdering, err
+	}
+
+	var cleanImageKeys []string
+	//clean out the keys, we don't want the yaml interfering with the yaml :P
 	for _, v := range imageKeys {
-		imageUrl := a.site.GetPhotoForKey(v)
-		imageUrls = append(imageUrls, imageUrl)
+		if strings.HasSuffix(v, ORDERING_YAML_NAME) {
+			//for now, just do nothing, we simply want to avoid appending,
+			//when we agree on a list of valid formats, we can ditch this check.
+		} else {
+			cleanImageKeys = append(cleanImageKeys, v)
+		}
+	}
+
+	//okay, now we're ready for processing and merging.
+	//some ground rules:
+	//0) if there is no ordering file, or an error retrieving/parsing the file, everything must work as
+	//   if there was never an ordering file in the first place.
+	//1) if an image is in the config but not in the bucket, it's silently dropped
+	//2) ordering of keys in config come before ordering of keys in bucket (i.e: listed in config THEN non-listed)
+	//3) each section is independent and optional but has some interlinkages, (this gets difficult because you
+	//   don't want to pick out the same photo for both cover and thumbnail.
+
+	//let's start with the cover photo, there's only one, this should be easy.
+
+	if albumOrderingConfig.Cover != "" {
+
+		// not the most efficient way of checking for existence, but it's one-off so not worth
+		// building a map/set.
+		var coverKeyInBucket = false
+		for _, bucketKey := range cleanImageKeys {
+			if strings.TrimLeft(bucketKey, "/") == strings.TrimLeft(albumOrderingConfig.Cover, "/") {
+				coverKeyInBucket = true
+				break
+			}
+		}
+
+		if coverKeyInBucket {
+			albumOrdering.Cover = a.site.GetPhotoForKey(albumOrderingConfig.Cover)
+		} else {
+			fmt.Printf("\ncover photo specified in ordering file not found in bucket, check %s exists. "+
+				"Falling back to first photo", albumOrderingConfig.Cover)
+			if len(cleanImageKeys) > 0 {
+				albumOrdering.Cover = a.site.GetPhotoForKey(cleanImageKeys[0])
+			} else {
+				albumOrdering.Cover = a.site.GetPhotoForKey("")
+			}
+		}
+	} else {
+		if len(cleanImageKeys) > 0 {
+			albumOrdering.Cover = a.site.GetPhotoForKey(cleanImageKeys[0])
+		} else {
+			albumOrdering.Cover = a.site.GetPhotoForKey("")
+		}
+	}
+
+	//thumbnails - there is a bit of duplicate code here, but it was clearer to do it
+	//this way rather than to reduce code and be opaque
+	var thumbKeys []string
+	if len(albumOrderingConfig.Thumbnails) > 0 {
+		thumbKeys = mergeList(cleanImageKeys, albumOrderingConfig.Thumbnails, a.Path)
+		if len(thumbKeys) > 5 {
+			thumbKeys = thumbKeys[0:5]
+		} else if len(thumbKeys) > 0 {
+			thumbKeys = thumbKeys[0:]
+		}
+	} else {
+		//take care of the offset here (i.e: cover is index 0 if we're not using the config)
+		thumbKeys = make([]string, len(cleanImageKeys))
+		copy(thumbKeys, cleanImageKeys)
+		if len(thumbKeys) > 6 {
```

Comment:
Instead of the `copy` call, maybe just use the `cleanImageKeys` here in the slice assignment below? Unless I'm missing something?

Reply:
the main reason (even though it doesn't actually happen here) was to avoid confusing side-effects. If we do `thumbKeys = cleanImageKeys[1:6]` (or something similar where we're just pointing to a slice of the same object in memory*), any later modification of thumbKeys (which again, we don't happen to do right now) could lead to unexpected orderings in `cleanImageKeys` (the side effect being that the actual album ordering is affected by a change in ordering of thumbKeys).

Now it's unlikely that anybody would want to change `thumbKeys` in the future, but i prefer the clarity here. If you don't like it I'd be happy to change it, if i misunderstood and you meant something other than 

All that said, while investigating how to do this safely without a copy, I've come across a good refactor, which I'll update shortly. 

I think I'll be finding a way to copy a smaller subset, would that work for you?

\* this is what i've understood through some simple experimentation with golang: https://play.golang.com/p/3B4qQVpCdI